library/src/scala/IArray.scala

```diff
@@ -153,165 +161,167 @@ object opaques
 
     /** Splits this array into a prefix/suffix pair according to a predicate. */
     def [T](arr: IArray[T]) span(p: T => Boolean): (IArray[T], IArray[T]) =
-      genericArrayOps(arr).span(p) match {
-        case (x, y) => (x.asInstanceOf[IArray[T]], y.asInstanceOf[IArray[T]])
-      }
+      genericArrayOps(arr).span(p)
 
     /** Splits this array into two at a given position. */
     def [T](arr: IArray[T]) splitAt(n: Int): (IArray[T], IArray[T]) =
-      genericArrayOps(arr).splitAt(n) match {
-        case (x, y) => (x.asInstanceOf[IArray[T]], y.asInstanceOf[IArray[T]])
-      }
+      genericArrayOps(arr).splitAt(n)
 
     /** Tests whether this array starts with the given array. */
     def [T, U >: T: ClassTag](arr: IArray[T]) startsWith(that: IArray[U], offset: Int = 0): Boolean =
-      genericArrayOps(arr).startsWith(that.asInstanceOf[Array[U]])
+      genericArrayOps(arr).startsWith(that)
 
     /** The rest of the array without its first element. */
     def [T](arr: IArray[T]) tail: IArray[T] =
-      genericArrayOps(arr).tail.asInstanceOf[IArray[T]]
+      genericArrayOps(arr).tail
 
     /** An array containing the first `n` elements of this array. */
     def [T](arr: IArray[T]) take(n: Int): IArray[T] =
-      genericArrayOps(arr).take(n).asInstanceOf[IArray[T]]
+      genericArrayOps(arr).take(n)
 
     /** An array containing the last `n` elements of this array. */
     def [T](arr: IArray[T]) takeRight(n: Int): IArray[T] =
-      genericArrayOps(arr).takeRight(n).asInstanceOf[IArray[T]]
+      genericArrayOps(arr).takeRight(n)
 
     /** Takes longest prefix of elements that satisfy a predicate. */
     def [T](arr: IArray[T]) takeWhile(p: T => Boolean): IArray[T] =
-      genericArrayOps(arr).takeWhile(p).asInstanceOf[IArray[T]]
+      genericArrayOps(arr).takeWhile(p)
 
     /** Converts an array of pairs into an array of first elements and an array of second elements. */
     def [U: ClassTag, V: ClassTag](arr: IArray[(U, V)]) unzip: (IArray[U], IArray[V]) =
-      genericArrayOps(arr).unzip match {
-        case (x, y) => (x.asInstanceOf[IArray[U]], y.asInstanceOf[IArray[V]])
-      }
+      genericArrayOps(arr).unzip
 
     /** Returns an array formed from this array and another iterable collection
       * by combining corresponding elements in pairs.
       * If one of the two collections is longer than the other, its remaining elements are ignored. */
     def [T, U: ClassTag](arr: IArray[T]) zip(that: IArray[U]): IArray[(T, U)] =
-      genericArrayOps(arr).zip(that).asInstanceOf[IArray[(T, U)]]
+      genericArrayOps(arr).zip(that)
   }
 end opaques
 
 type IArray[+T] = opaques.IArray[T]
 
 object IArray {
+  import opaques.Sub
+  import opaques.Sup
+
+  // It is not possible use a private Conversion instance as it leaks through `inline` methods;
+  // see: https://github.com/lampepfl/dotty/issues/8194
+  private inline implicit def lift[A](arr: Array[A]): IArray[A] = arr: Sub[A]
```

Comment:
Did you see my original comment?

> I used a private inline implicit conversion for convenience in the `IArray` object. I tried to use an instance of `Conversion` instead, but that did not work because of #8194.

Is there a way to make a `Conversion` completely inline, so it disappears from the method bodies?